src/webui/www/private/scripts/piecesbar.js

```diff
@@ -0,0 +1,266 @@
+                'height': 0,
+                'downloadingColor': 'green',
+                'haveColor': 'blue',
+                'borderSize': 1,
+                'borderColor': '#999'
+            };
+
+            if (parameters && $type(parameters) == 'object')
+                $extend(vals, parameters);
+            vals.height = Math.max(vals.height, 12);
+
+            const obj = new Element('div', {
+                'id': vals.id,
+                'class': 'piecesbarWrapper',
+                'styles': {
+                    'border': vals.borderSize + 'px solid ' + vals.borderColor,
+                    'height': vals.height,
+                }
+            });
+            obj.vals = vals;
+            obj.vals.pieces = $pick(pieces, []);
+
+            obj.vals.canvas = new Element('canvas', {
+                'id': vals.id + '_canvas',
+                'class': 'piecesbarCanvas',
+                'width': vals.width - (2 * vals.borderSize),
+                'height': 1
+            });
+            obj.appendChild(obj.vals.canvas);
+
+            obj.setPieces = setPieces;
+            obj.refresh = refresh;
+            obj.clear = setPieces.bind(obj, []);
+            obj._drawStatus = drawStatus;
+
+            if (vals.width)
+                obj.setPieces(vals.pieces);
+            else
+                setTimeout(() => { checkForParent(obj.id); }, 1);
+
+            return obj;
+        }
+    });
+
+    function setPieces(pieces) {
+        if (!Array.isArray(pieces))
+            pieces = [];
+
+        this.vals.pieces = pieces;
+        this.refresh(true);
+    }
+
+    function refresh(force) {
+        if (!this.parentNode)
+            return;
+
+        const width = this.offsetWidth;
+        if (this.vals.width === width && !force)
+            return;
+
+        this.vals.width = width;
+
+        // change canvas size to fit exactly in the space
+        this.vals.canvas.width = width - (2 * this.vals.borderSize);
+
+        const canvas = this.vals.canvas;
+        const ctx = canvas.getContext('2d');
+        ctx.clearRect(0, 0, canvas.width, canvas.height);
+
+        const pieces = this.vals.pieces;
+        const imageWidth = canvas.width;
+
+        if (pieces.length === 0)
+            return;
+
+        let minStatus = Infinity;
+        let maxStatus = 0;
+
+        for (const status of pieces) {
+            if (status > maxStatus) {
+                maxStatus = status;
+            }
+
+            if (status < minStatus) {
+                minStatus = status;
+            }
+        }
+
+        // if no progress then don't do anything
+        if (maxStatus === 0)
+            return;
+
+        // if all pieces are downloaded, fill entire image at once
+        if (minStatus === STATUS_DOWNLOADED) {
+            ctx.fillStyle = this.vals.haveColor;
+            ctx.fillRect(0, 0, canvas.width, canvas.height);
+            return;
+        }
+
+        /* Linear transformation from pieces to pixels.
```

Comment:
For the very wide with few pieces case, I did modify it a bit so that it only draws a rectangle when the color is different. So if 30px are all one color it will only draw the 30px once it sees a change in the color.

But it still makes sense, for the maximum size of the image use either real width in pixel or number of pieces, whichever is smaller. I will update that.

For the issue of maximum canvas size, that is very interesting and I didn't know there was a maximum size, thanks for looking into that! I'm not so sure that using such a large SVG will perform much better, since I believe that would mean manipulating many DOM nodes which I think would be slower than adding pixels to a canvas, and the issue only shows up with very wide screens. But I could change it so that the maximum width is 4096 and that should satisfy all cases about max size. So for anyone who has a screen large enough to reach that maximum it would be ok to show only 4096 pixels and stretch it as needed. It should still provide a good level of detail to be useful. Any thoughts on that?